### PR TITLE
Issue-423: Crash when editing device dataitem in DeviceModel

### PR DESCRIPTION
**Pull request by mnoomnoo**
#423 

Updated Checkpoint::updateDataItems() to remove orphans from observations map. Updated CircularBuffer::updateDataItems() to remove orphans from m_slidingBuffer

### CODE REVIEW

File: src/mtconnect/buffer/circular_buffer.hpp [hunk 1]
```diff
@@ -82,11 +85,29 @@ namespace mtconnect::buffer {
...
+        else {
+          o->updateDataItem(diMap);
+        }
+        iter++;
+      }
+
+      // remove orphans from the slidingBuffer
+      while(!orphanBufferItems.empty()) {
+        auto orphanIterIter = orphanBufferItems.back();
+        m_slidingBuffer.erase(orphanIterIter);
```
Comment (wsobel) [inline src/mtconnect/buffer/circular_buffer.hpp:106]: We can't erase from the sliding buffer. The issue is it messes up the sequence number computation. I'll check into this and see if I can work around the issue. It should be sufficient to have them orphaned. Did you try without this code?

**Review by wsobel — CHANGES_REQUESTED** [review]

**Comment by wsobel** [comment]
We can't erase from the circular buffer since the positions are computed using the sequence numbers. it is better to leave holes. When we get observations from the circular buffer, it skips orphaned observations. The checkpoint code looks correct. 

Did you write a test to verify the changes?

File: src/mtconnect/buffer/circular_buffer.hpp (hunk 1 above)
Comment (mnoomnoo) [reply src/mtconnect/buffer/circular_buffer.hpp:106]: It appears to work without erasing the orphaned dataItems from the sliding buffer. I'm testing my changes and will commit them soon.

**Comment by mnoomnoo** [comment]
I updated CircularBuffer::updateDataItems() to skip orphaned dataItems instead of deleting them from the sliding buffer. Checkpoint::updateDataItems() still deletes orphaned observations.

> Did you write a test to verify the changes?

I didn't write a test to verify the changes. I can take a peek at existing unit tests to see if there is something I can modify to test the agent when data items are added / removed or their IDs are changed.

**Comment by wsobel** [comment]
Thanks a lot. I will review changes later today. - W

**Review by wsobel — APPROVED** [review]

**Pull request merged**